packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/queries.ts

```diff
@@ -244,72 +244,123 @@ LEFT JOIN owners_total_utxos otu on
-    THEN (
-      COALESCE((rewards.total_amount - (rewards.total_amount + update.fixed_cost))
-      /total_stake.active_stake, 0)
-    )
-    ELSE (
-      COALESCE((rewards.total_amount - (((rewards.total_amount-update.fixed_cost) * update.margin) + update.fixed_cost))
-      /total_stake.active_stake, 0)
-    ) END AS member_roi
-  FROM total_rewards_per_epoch rewards
-  JOIN epochs e on 
-    e."no" = rewards.epoch_no
-  LEFT JOIN total_stake_per_epoch total_stake on 
-    rewards.epoch_no = total_stake.epoch_no
-  JOIN pool_update update
-    ON update.id = (
+pool_rewards_per_epoch AS (
+	SELECT
+		reward.pool_id AS hash_id,
+		epochs.epoch_no,
+		SUM(reward.amount) AS total_amount
+	FROM epochs
+	JOIN reward 
+		ON reward.earned_epoch = epochs.epoch_no
+		AND reward.pool_id = ANY($1)
+	GROUP BY reward.pool_id, epochs.epoch_no
+),
+pool_stake_per_epoch AS (
+	SELECT
+		epoch_stake.pool_id AS hash_id,
+		epochs.epoch_no,
+    SUM(epoch_stake.amount) AS active_stake
+	FROM epochs
+	JOIN epoch_stake 
+		ON epoch_stake.epoch_no = epochs.epoch_no
+		AND epoch_stake.pool_id = ANY($1)
+	GROUP BY epoch_stake.pool_id, epochs.epoch_no
+),
+epoch_rewards AS (
+	SELECT 
+		epochs.epoch_no,
+		epochs.epoch_length,
+		stake.hash_id,
+		COALESCE(rewards.total_amount, 0) AS total_rewards,
+		COALESCE(stake.active_stake, 0) AS active_stake,		
+		CASE 
+			WHEN pool.fixed_cost >= rewards.total_amount
+	    	THEN COALESCE(rewards.total_amount, 0)
+	    	ELSE (
+	      	COALESCE(
+	      		FLOOR((rewards.total_amount - pool.fixed_cost) * pool.margin) + pool.fixed_cost
+	      	, 0)
+	    	) 
+		END AS operator_fees,
+		CASE 
+			WHEN COALESCE(stake.active_stake, 0) = 0
+				THEN 0
+			WHEN pool.fixed_cost >= rewards.total_amount
+		    THEN (
+		      COALESCE(
+		      	(rewards.total_amount - COALESCE(rewards.total_amount, 0)) / stake.active_stake
+		      , 0)
+		    )
+		    ELSE (
+		      COALESCE(
+						(rewards.total_amount - 
+							COALESCE(
+	      				FLOOR((rewards.total_amount - pool.fixed_cost) * pool.margin) + pool.fixed_cost
+	      			, 0)) / stake.active_stake
+		      , 0)
+		    ) END AS member_roi
+  FROM pool_stake_per_epoch AS stake 
+  JOIN epochs 
+  	ON epochs.epoch_no = stake.epoch_no
+  LEFT JOIN pool_rewards_per_epoch AS rewards
+  	ON rewards.epoch_no = stake.epoch_no
+  	AND rewards.hash_id = stake.hash_id
+  JOIN pool_update AS pool
+    ON pool.id = (
       SELECT id
-      FROM pool_update pu2
-      WHERE pu2.hash_id = $1 
-        and pu2.active_epoch_no <= rewards.epoch_no
+      FROM pool_update
+      WHERE hash_id = stake.hash_id 
+        AND active_epoch_no <= epochs.epoch_no
       ORDER BY id DESC
       LIMIT 1
     )
-   order by rewards.epoch_no desc
+)`;
+
+export const findPoolEpochRewards = (limit?: number) => `
+  ${epochRewardsSubqueries(limit)}
+  SELECT * FROM epoch_rewards
+  ORDER BY epoch_no desc
+`;
+
+export const findPoolAPY = (limit?: number) => `
```

Comment:
I'm getting some very big APY values for some stake pools, this was based on `packages/core/src/Cardano/util/estimateStakePoolAPY.ts`.
Are we sure the formulas for this and `memberROI` are correct?

Reply:
In my local db with testnet data there are 67 stake pools with APY >= 1. The biggest one is
```
hash_id|apy                  |
-------|---------------------|
    226|   871048215088976000|
```

Epoch rewards for that stake pool are:
```
epoch_no|epoch_length|hash_id|total_rewards|active_stake|operator_fees|member_roi        |
--------|------------|-------|-------------|------------|-------------|------------------|
      89|   431989000|    226|            0|   693587479|          0.0|               0.0|
      88|   431983000|    226|    500000000|    97247840|    352000000|1.5218847020149753|
```

Avg. daily ROI:
```
hash_id|avg_roi            |
-------|-------------------|
    226|0.15219445933531395|
```

Avg. daily ROI calculation in the code:
```ts
const roisPerDay = rewardsHistory.map(
    ({ epochLength, memberROI }) => memberROI / (epochLength / MILLISECONDS_PER_DAY)
  );
const averageDailyROI = sum(roisPerDay) / roisPerDay.length;
```
which would be
``` 
(0 + (1.5218847020149753 / (431983000/86400000)) / 2 = 
0.30438891867 / 2 = 
0.15219445933
```

And then to calculate the final number:
```ts
const epochLengthInDays = rewardsHistory[rewardsHistory.length - 1].epochLength / MILLISECONDS_PER_DAY;
// epochLengthInDays = 431989000 / 86400000 = 4.99987268519

const roiPerEpoch = averageDailyROI * epochLengthInDays;
// roiPerEpoch = 0.15219445933 * 4.99987268519 = 0.76095292004

const numEpochs = 365 / epochLengthInDays;
// numEpochs = 365 / 4.99987268519 = 73.0018588436

return Math.pow(1 + roiPerEpoch, numEpochs) - 1;
// APY = ((1 + 0.76095292004) ^ 73.0018588436) - 1 = 8.7104821e+17
```

which is what I'm getting from the query

Reply:
For the member roi calculation, in epoch 88 the total rewards (500000000) are greater than fixed cost (340000000) so it does:
```
(rewards - (FLOOR((rewards - cost) * margin) + cost)) / stake =

(500000000 - (((500000000 - 340000000) * 0.075) + 340000000)) / 97247840 =

1.52188470201
```

in epoch 89 fixed cost is greater than the rewards (0), and it does:
```
(rewards - rewards) / stake
```
which will always return 0

Reply:
Another things is that rewards can be type `[leader, member, reserves, treasury, refund]`, is it ok that we take all of them into account when calculating the roi and stuff?

Reply:
`1.5218847020149753` for `member_roi` is way too much, it would be 152% interest per epoch :laughing: Could be possible with very small stake and incredible luck though? :thinking: 

> Another thing is that rewards can be type [leader, member, reserves, treasury, refund], is it ok that we take all of them into account when calculating the roi and stuff?

No, we're only interested in member rewards here when computing APY, so something like:
```
epoch_member_rewards = SUM(SELECT ... WHERE epoch=... AND type="member")
member_roi = epoch_member_rewards / active_stake
```

Reply:
With `member` rewards only the numbers make more sense, biggest member roi being `0.0002945421377973` and APY `0.012573814334454259`

Reply:
Hmm that's much better, but now it seems too small for the biggest member roi. Although I guess it could be correct for the testnet data :thinking Are you using somewhat fully synced data? We might want to test this on an up-to-date one.

Reply:
I stopped db-sync after restoring the snapshot at block 3590461

Reply:
I'm getting 0 roi and apy for this one https://testnet.cexplorer.io/pool/pool167u07rzwu6dr40hx2pr4vh592vxp4zen9ct2p3h84wzqzv6fkgv

Reply:
If you look at https://testnet.cexplorer.io/pool/pool167u07rzwu6dr40hx2pr4vh592vxp4zen9ct2p3h84wzqzv6fkgv/rewards#data delegators get negligible rewards. I don't think we should be counting in operator (`leader`) rewards when computing APY

Reply:
Maybe missing a cast to float somewhere so it gets to 0 :thinking:

Reply:
No, it's 0 because of this case at the member roi calculation: 
```sql
   WHEN pool.fixed_cost >= rewards.total_amount
    THEN (
      --  (rewards - rewards) / stake will always be 0
      COALESCE(
      	(rewards.total_amount - COALESCE(rewards.total_amount, 0)) / stake.active_stake
      , 0)
    )
```
This stake pool rewards are less than the fixed_cost
```
fixed_cost|total_rewards|
----------|-------------|
 340000000|     33436585|
 340000000|     31308365|
 340000000|     29652223|
 340000000|     29274580|
 340000000|     28246202|
 340000000|     27212254|
```

Reply:
https://testnet.cardanoscan.io/pool/d7b8ff0c4ee69a3abee65047565e85530c1a8b332e16a0c6e7ab8401 also shows lifetime rewards as 0. Seems like we're good then?